### ISSUE TITLE
[BUG]_cat/nodes return stale OpenSearch version after upgrades from ES 7.x/6.8

### ISSUE THREAD

**Issue by Gaurav614** [opensearch-project/OpenSearch#827]
**Describe the bug**
With the BWC fix, done as part of this [PR](https://github.com/opensearch-project/OpenSearch/commit/26d579287f50bb33e17c8fe1f05ea208d5c64d1f), the upgrades were happening smoothly. After the upgrades finishing, the _cat/nodes still return the `7.10.2` response as version even after the master switch to OS-1.0.0. This is gives a false information since the version is upgraded to OS-1.0.0.

**To Reproduce**
1. Create a cluster with 3 dedicated master node and 3 data nodes cluster.
2. Do upgrade to Opensearch 1.0.0
3. after upgrade try this hitting  _cat/nodes api
```
curl localhost:9200/_cat/nodes?h=node.role,version,master
imr 1.0.0  -
imr 1.0.0  -
imr 1.0.0  *
di  7.10.2 -
di  7.10.2 -
di  7.10.2 -
```

We need to then kill the process to get the current response from the _cat/nodes. 
**Expected behavior**
```
curl localhost:9200/_cat/nodes?h=node.role,version,master
imr 1.0.0  -
imr 1.0.0  -
imr 1.0.0  *
di  1.0.0 -
di  1.0.0 -
di  1.0.0 -
```

**Comment by shwetathareja** [comment]
the current hypothesis on why this happening

As part of fixing version compatibility with Elasticsearch nodes during upgrade (https://github.com/opensearch-project/OpenSearch/pull/775). The BWC version (7.10.2) is passed from 2 places 1. during transport handshake 2. discovery node -  the version is used when node is joining the master. Consider the below flow:
1. There is an ES cluster (6.8 <= version <= 7.10.2)
2. Now OpenSearch nodes join the cluster. As master is an ES node, they pass their version as 7.10.2 to master and it gets stored in in-memory cluster state (DiscoveryNodes).
3. Upgrade is finished and one of the OpenSearch master eligible node becomes the master.
4. At this point, as all nodes are OpenSearch, it is expected all APIs including _cat/nodes should return the version as 1.0.0
5. But this is not the case as reported above, This is due to the reason that when OpenSearch node became the master, it had the cluster state from ES node which was the master and had the version as 7.10.2 in the state for OpenSearch nodes as well. And the _cat/nodes API response is generated from cluster state.

The current workaround is to bounce the nodes and they will reconnect to the master and version will show up as 1.0.0 until this issue is fixed.

**Comment by shwetathareja** [comment]
@nknize @dblock @itiyamas : Need your opinion on this - 
While debugging this found another issue:

When Elasticsearch node is master (during upgrade), it returns the version as below

Elasticsearch is the master and rest call comes to Elasticsearch Node
```
sh-4.4# curl localhost:9200/_cat/nodes?h=ip,id,name,version,master
172.26.0.2 roLY opensearch-1    7.10.2 -
172.26.0.3 Ez1X opensearch-3    7.10.2 -
172.26.0.4 AWIm elasticsearch-2 7.9.0  *
```

Elasticsearch is the master and rest call comes to OpenSearch Node
```
sh-4.4# curl localhost:9200/_cat/nodes?h=ip,id,name,version,master
172.26.0.2 roLY opensearch-1    7.10.2 -
172.26.0.3 Ez1X opensearch-3    7.10.2 -
172.26.0.4 AWIm elasticsearch-2 7.9.0  *
```

It will be full BWC mode.

Now, consider OpenSearch node has become master, should we let Elasticsearch nodes connect to OpenSearch master or not. In order to let Elasticsearch node continue to connect to it, OpenSearch master will have to send BWC version (7.10.2) to ES nodes for all OpenSearch nodes in the cluster and API responses will look like below:

OpenSearch is the master and rest call comes to any OpenSearch node:
```
sh-4.4# curl localhost:9200/_cat/nodes?h=ip,id,name,version,master
172.26.0.2 roLY opensearch-1 1.0.0 *
172.26.0.3 Ez1X opensearch-3 1.0.0 -
172.26.0.4 AWIm elasticsearch-2 7.9.0 -
```

OpenSearch is the master and rest call comes to Elasticsearch Node
```
sh-4.4# curl localhost:9200/_cat/nodes?h=ip,id,name,version,master
172.26.0.2 roLY opensearch-1 7.10.2 *
172.26.0.3 Ez1X opensearch-3 7.10.2 -
172.26.0.4 AWIm elasticsearch-2 7.9.0 -
```

If the above APIs are ok, then we will need to call out explicitly, so that this doesn't cause any confusion. Other option is to prevent Elasticsearch node from connecting to OpenSearch node altogether which could be more disruptive but no such confusion. I want all of your opinion on this.

**Comment by dblock** [comment]
I agree that it would be disruptive if ES nodes can't connect to OpenSearch master. That feels like a broken system.

Are you proposing that `_cat/nodes` return different things depending on who you ask? That seems odd to me, as you already point out. In bwc mode, what are the reasons not to always return the bwc version across the board, then turn the mode off once everything is rolled over to OpenSearch?

**Comment by shwetathareja** [comment]
Thanks @dblock for taking a look.

> I agree that it would be disruptive if ES nodes can't connect to OpenSearch master. That feels like a broken system.
> 
Normally, the expectation is that user upgrades the non master eligible nodes first and then master eligible nodes, so that in case of incompatibility there are no issues as much as possible. From Elasticsearch documentation as well
_"Upgrading the nodes in this order ensures that the master-ineligible nodes are always running a version at least as new as the master-eligible nodes. Newer nodes can always join a cluster with an older master, but older nodes cannot always join a cluster with a newer master."_

Having said that yes not allowing nodes to connect is very disruptive and should be the last resort.

> Are you proposing that `_cat/nodes` return different things depending on who you ask? That seems odd to me, as you already point out, but not necessarily bad.
> 
> In bwc mode, what are the reasons not to always return the bwc version across the board, then turn the mode off once everything is rolled over to OpenSearch?

The issue is that there is no definite way to say everything is rolled over to OpenSearch as customer can choose to use rolling upgrade or B/G deployment. There is an option to identify if the master is OpenSearch node and there are any Elasticsearch nodes in the cluster or not. In case there are no Elasticsearch nodes, the OpenSearch node can switch off the BWC mode but this is not reliable as Elasticsearch nodes could disconnect temporarily and join back. At this point, we end up with 2 options:
1. Return different responses from _cat/nodes, this ensure that Elasticsearch nodes can still join the cluster
2. Or, prevent Elasticsearch nodes from joining the cluster altogether.

**Comment by dblock** [comment]
I see the problem @shwetathareja. Thinking of (2), is there a failure case in a network partition where master thinks it has upgraded, prevents all ES nodes from joining, then another node re-appears and can no longer connect? Maybe there should be an actual API that says "starting upgrade" (which does all the things that we document such as flushing data, etc.) and "I'm done upgrading". The latter would do (2).

I think for 1.0 I'd just continue doing (1).

**Comment by shwetathareja** [comment]
> I see the problem @shwetathareja. Thinking of (2), is there a failure case in a network partition where master thinks it has upgraded, prevents all ES nodes from joining, then another node re-appears and can no longer connect? Maybe there should be an actual API that says "starting upgrade" (which does all the things that we document such as flushing data, etc.) and "I'm done upgrading". The latter would do (2).
> 
Yes, not just during network partition, consider few data nodes running high on JVM and got disconnected and master switched to OpenSearch node and now these data will not be able to connect to OpenSearch node (2). Yeah, we can think of providing an upgrade API. The current work around the user can do is to put the OpenSearch masters in voting config exclusion, that ensures they never become the master until the exclusion is removed. This is an external control which can be used by the user during upgrade.  

> I think for 1.0, short of an upgrading API, I'd just continue doing (1).
> 
> I think that refreshing nodes after OpenSearch becomes master is OK and the network partitioning scenario can be ignored because in that case the node would not be master, and you would just upgrade that node that came back, and call it a day.

The current fix in Draft PR is to refresh the version as soon as OpenSearch node becomes master. Also the confusion in API response only arises if that Elasticsearch node that has come up will serve actual requests.

**Comment by nknize** [comment]
fixed by #865
